### ISSUE TITLE
Support Dependency Injection in places where concrete implementations are used

### ISSUE THREAD

**Issue by patrickhuber** [patrickhuber/Pliant#60]
In order to provide a more decoupled architecture, support a platform agnostic form of dependency injection for instances where concrete classes are used.

**Comment by bilsaboob** [comment]
Hey,

While working on the "grammar workbench" I could do some refactoring to "decouple" some of the dependency if you would like?

I'm not sure if you have any concrete thoughts on how you would like to proceed in that case, but from Tooling perspective I think that the following structure could work ok.

Pliant
- responsible for "util" tools and any "infrastructure" code to support the Pliant library... Extension methods and other convenience helpers...

* Pliant.Lib
- responsible for exposing "model" classes working with defining a Grammar and any related code.
- the purpose is to keep this project as aganostic of the "runtime" implementation, such as the inner workings of the Earley algorithm and any other details that you as library author would like to be able to change "on demand" without breaking the full tooling chain and community extensions.
- will contain classes related to:
  * Grammar/Rule interfaces - anything need to define a Grammar
  * Bootstrapper for the "Parsing engine" - that will expect certain "dependencies" to be injected when bootstrapping the "selected" Pliant.Runtime library / reference...
  * Any "parse forest" related interfaces, no concrete implementations - and nothing that reference the "runtime logic"

* Pliant.Lib.Grammar
- responsible for handling anything regarding "building a grammar", while the Pliant.Lib contains the grammar interfaces but doesn't care about how a grammar is constructed.
- will contain classes related to:
  * GrammarBuilder - "helper" that is responsible for building a "Grammar implementation"
  
* Pliant.Lib.AST
- responsible for AST base classes / visitors - building on top of the already exiting Visitor concept in Pliant, except that it "automates" the handling of "ParseForest nodes" into a more "simple" way of resolving - something similar to how Antlr does it.
- will contain classess related to:
  * AST base classes etc...
  * AST Visitors and any helper tools for analyzing an AST

* Pliant.Runtime
- responsible for the actual Earley algorithm, any way you as author find necessary - you can do any change you would like here without really thinking about the "end user" and what libraries he references.
- will contain classes related to:
  * Earley parsing algorithm / NFA and DFA / etc...
  * Concrete implementation / overrides of anything from the Pliant library required by the "framework"
- will not really be referenced by "community projects" - unless they need/want access to "internals" of the algorithms...



This structure will be used according to:

* Grammar community projects:
- These will reference the Pliant.Lib / Pliant.Lib.Grammar and generally use the "Grammar" building blocks, will not need the Runtime.

* End user projects that want to use an already developed grammar:
- These projects will reference the Pliant.Lib / "Pliant.CSharpGrammar" (or any target grammar) and any version of the Pliant.Runtime (given that it's compatible with the Pliant.Lib)

* Community tools for building grammars:
- These will reference the Pliant.Lib / Pliant.Lib.Grammar and not really care about the Runtime

* Community "core runtime" extensions - such as "customized parts" of the Earley algorithm (custom error recovery strategies? / Custom lex implementations etc...)
- These will reference the Pliant.Lib / Pliant.Lib.Grammar and most likely the Pliant.Runtime as well.


If you agree on this, I'll start pulling out things into this specific structure - if you rather do it yourself at some point - that ok :) - just a thought, since I was thinking of starting on the "grammar workbench" anyway... and I'll get a feel for what I would need from the Pliant at that point anyway.

If you would like me to do this - then on to the next question - how do we sync the work? - my thoughts are that I would like to have "free reign" while doing this, and then let you pick whatever pieces you want and do any renaming refactoring you feel necessary? 
- So perhaps I'll make a fork and work on that, and make a pull request to your repo?
- So basically I would work with the following structure:
  * Fork repository of Pliant
    * forked master (I'm not really sure I'm interested in this one...)
    * forked development
      - this is where I'll do the development and possibly make some subbranch at some point if I need
      - I would make a pull request back to your "development" branch? as well ass getting the latest stuff from your development branch?

Any thoughts?

**Comment by bilsaboob** [comment]
Btw, do you have any preferred libraries for: (I'll be using this for the grammar workbench - I guess it would be useful to use the same in case you want it in the Pliant libs at some point)

* DI - Autofac perhaps?
* Logging - NLog?
* Persistence? - I'm currently using a "document DB" ala MongoDB for the "old" grammar workbench... basically for storing "changes" and history and such...
* GUI - WPF? (in combination with MVVMCross?)

**Comment by patrickhuber** [comment]
I have a similar approach in mind. I tried to keep things in isolated folders so those are the first level of grouping. I'll lay out how things are currently structured and we can see what the lines are and what assembly it makes sense to push them into in order to achieve some feature by consumers. 

Here is the dependency graph:
![pliant-dependencies](https://cloud.githubusercontent.com/assets/182398/18181706/8302c890-705a-11e6-860e-d83e0b12161d.png)

(I'll edit this if you see something I missed)

* Pliant.Ebnf - pull out into its own project. You won't need this unless you want a EBNF DSL. 
* Pliant.Bnf  - pull out into its own project. You won't need this unless you want a BNF DSL. 
* Pliant.Common - core components can be pulled into a single assembly. I've also heard this called the "core" classes. Pliant.Core
    * Pliant.Collections
    * Pliant.Utilities
    * Pliant.Diagnostics
* Pliant.Reactive  - delete, I was going to use this for some event generation but never did. 
* Pliant.Grammars - pull out into its own project
    *  this has a dependency on TokenType, so maybe roll TokenType into the Grammar namespace?
* Pliant.RegularExpressions - pull out into its own project. This currently has a circular dependency on the parse engine. I guess I didn't think that one out. I imagine the logic isn't unique, it could probably be pulled into a Parser class for the Runtime project. 
    * References:
        * Pliant.Automata
* Pliant.Builders - pull into its own project. Maybe call it Pliant.Grammar.Builders? 
* Pliant.Ast ( I never capitalize 3 letter acronyms because of this: https://msdn.microsoft.com/en-us/library/ms229043(v=vs.110).aspx) - Break out into its own assembly. I like keeping the Forest and Tree in their own folders. If it makes sense these could be broken out into two individual projects as well. 
* Pliant.Charts - combine with Runtime but keep in a subfolder? is very tightly coupled to the Pliant.Runtime llibrary and holds a lot of core Earley logic. I think it makes sense to keep Earley logic in the same assembly. The big problem with moving this away from the Forest folder will be the parse nodes have references to states and vice versa. I created a IParseNode interface to help with that but haven't gone through pushing the dependency to be one way. 
* Pliant.Automata - pull into its own assembly (these classes could probably be made more generic than using terminals directly)
    * References : 
        * Pliant.Grammars
        * Pliant.Core / Pliant.Common
        * currently there is a reference to Pliant.Tokens but that is TokenType which can resolve to Pliant.Grammars
* Pliant.Runtime - pull into its own assembly. Maybe look at renaming to Pliant.Earley or something if I ever get around to implementing a GLL or GLR parser. 
    
If you aren't partial to the "Lib" suffix, I'd like to keep it out in favor of just Pliant.*. The runtime folder was just to keep IParseEngine and its supporting classes out of the project root, but it is still just a library. 

For contributions, yes a fork of the development branch is probably the best place to start. In other projects I've seen them create pull request very early so that any changesets are updated and we can go over the changes. You or I can make comments before the pull request is accepted. The nice thing about that is the changes are added to the pull request automatically. That way its not a big huge pull request that takes forever to review and we can incrementally improve the code before submission. It also gives me enough time to update the CI build and CD pipeline for the NuGet package / packages. 

For the DI I usually don't directly reference a DI framework and instead create a interface called IDependencyResolver. 

So as a start:

```csharp
public namespace Pliant.Ioc
{
   public interface IDependencyResolver
   {
         T GetService<T>();
         T GetService<T>(string name);
         object GetService(Type type);
         object GetService(Type type, string name);         
   }
}
```

Then, a consumer would create a binding library like Pliant.AutoFac or Pliant.Workbench.AutoFac and implement IDependencyResolver. That way we can reduce coupling and allow the user to specify the DI container of their choice. I sometimes will implement some additional Resolve  methods and wire up a simple dictionary for unit tests, etc. I used to do a lot of NHibernate and they had a similar philosophy: provide an interface and provide a default implementation. Allow the user to override. 

Logging, similar approach. Currently I use System.Diagnostics.Trace for the grammar execution and only log on Debug. There is a huge performance hit in doing logging during the core loop, so I had to make sure I wasn't doing any memory allocations or anything in there. If there as a ILogger interface in the Common/Core library, I could reference that instead and rely on DI to provide a NullLogger or something from one of the logging frameworks. 

For Database, I'm not sure I have a strict preference. I probably use sql server too much :). A similar plugin model to DI could be used. As long as the implementation of the persistence layer is abstract enough. Entity framework and System.Data.Common already do a pretty good job of abstraction, so as long as you code to one of those and try to keep implementation detail (DDL etc) out of the generic implementation you should be fine. 

I don't really have a preference on UI. I went the web track with my career. I've messed around with WPF, never with MVVM Cross. Is cross platform a concern? I've been thinking about a .NET core port but was waiting for the .net core tooling to hit GA.

**Comment by bilsaboob** [comment]
Hey, ok, sounds good to me :)

I'll give you a ping in the weekend once I'll start setting things up - perhaps we could stay in touch on Skype or so? - my profile name is "bilsaboob", feel free to add me!

Regards,
/Peter

**Comment by bilsaboob** [comment]
Hey,

I've started on setting up the projects / repos and such.

Currently I've setup the following structure:

* Pliant (I've called the fork as RapidPliant ... I'm also keeping a local Pliant repo for reference...)
 * in this fork I'm keeping the following branches:
   * development (I will try to keep this as up to date with your development branch)
    * rapid_development (this a branch of my development in the fork - where I will o the refactorings, and once "ready" I'll merge things down to the development - at which point we can discuss which changes and how to pull them)

* Pliant.GrammarWorkbench
 * this is a new repository, where I will be implementing the workbench application
 * it will reference the rapid_development branch directly during the refactoring work
   * at some point I'll want to remove the direct reference and replace it with a nuget reference or so...

* Pliant.Grammars.RapidBnf
 * this is a new repository for the Grammar for the "bnf syntax" that is to be used by the GrammarWorkbench (the one I'm currently having a handrolled parser for, hopefully this will give me better insight on the Pliant lib in general and what else may be "wanted" from IDE development perspective... incremental parsing / recovery and what not...)

* RapidSharp
 * this will be where I'll put my own personal hobby stuff... not sure if I will push everything into the repository, or perhaps just the "sample" where I'm using the Pliant grammar... or whenever discussion on something may arise...

You can view the repositories on my github.

Honestly I'm not entirely sure on how to handle the dependencies between github projects - my ideal situation would have been:
* GrammarWorkbench repo
  * references the following github projects as "references" in visual studio:
    * RapidPliant/rapid_development branch - whenever I make changes here... it also updates changes in the "local git repository of RapidPliant/rapid_development" branch...

* This would have allowed me to work in the "GrammarWorkbench" project and do all refactorings there, and also allow anyone that checks out the repository to get all required github project references...

* However the best I can think of is just to checkout the other repositories to a local directory on disk and make a relative path reference to those... it would still require the user to checkout those libraries...
  * In order to remedy I've had a look at Paket - https://fsprojects.github.io/Paket/ - it seems to have potential, currently I've only played around with "nuget" references, and I've tested the support to reference a nuget repository (which works fine) - however it always downloads the repository to the "paket-files" directory... and I can't find any documentation on this... (I'll have a look at the sourcecode and see if I can find it quickly)

One thing that came to mind is that perhaps we could use some better way for communication rather than writing here on the forum? - in order to follow up on changes / progress we are making? )

I would rather it not becoming a forced "jira" project or so :) - though perhaps have some screensharing / meeting every now and then on skype - walking through the change an such? (of course I don't want to make this a big overhead, basically just whenever someone feels the need for it... and we are able to fit it into our schedules...?

Regards,
/Peter

**Comment by patrickhuber** [comment]
I created a Gitter feed. From what I've seen its pretty common among the github community to do chat there. I'll end up putting the badge on the readme.md file. https://gitter.im/Pliant-net/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link. 

Haven't looked at Paket yet, I need to research it a bit more. I usually just use NuGet packages to share common libraries but that may not work with the workflow you are proposing. I keep going back to git submodules, but perhaps Paket will be the best option.

**Comment by bilsa** [comment]
Yep, i was thinking of subrepositories too - though i was mercurial - I'll hav a look at it for github too - if nothing else, paket seems like a tool to keep an eye on :) - i like the Idea of beeing able to have an intuitive config file - and to manage references to GitHub projects as if nuget packages ... anyway, I'll get back on this later in the gitter channel
